File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> frequencyMap;
        int n=nums.size();
        vector<int> ans;
        for(int i=0;i<n;i++){
            frequencyMap[nums[i]]++;
        }
        //Add elements in the heap with frequency, element as its sorted based on frequency
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
        for(auto x: frequencyMap){
            minHeap.push({x.second,x.first});
            if(minHeap.size()>k){
                minHeap.pop();
            }
        }
        while(!minHeap.empty()){
            ans.push_back(minHeap.top().second); 
            minHeap.pop();
        }
        return ans;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        PriorityQueue<Pair<Integer, Integer>> minHeap = new PriorityQueue<>((a,b)-> a.getKey() - b.getKey());
        Map<Integer, Integer> frequencyMap = new HashMap<>();
        for(int i=0;i<nums.length;i++){
            frequencyMap.put(nums[i], frequencyMap.getOrDefault(nums[i],0)+1);
        }
        for(Map.Entry<Integer, Integer> entry: frequencyMap.entrySet()){
            int frequency = entry.getValue();
            int element = entry.getKey();
            minHeap.add(new Pair<>(frequency, element));
            if(minHeap.size()>k){
                minHeap.poll();
            }
        }
        int result[] = new int[k];
        int i=0;
        while(!minHeap.isEmpty()){
            result[i++]=minHeap.poll().getValue();
        }
        return result;
    }
}